tasks/run_puppet_device.rb

```diff
@@ -16,6 +16,7 @@
 noop = (params['_noop']) ? '--noop' : ''
 target = (params['target']) ? params['target'] : ''
 timeout = (params['timeout'].to_i > 0) ? params['timeout'].to_i : default_timeout
+apply = (params['apply']) ? "--apply=#{params['apply']}" : ''
```

Comment:
Could you add a `File.file?(params['apply'])` test if `params['apply']` is set and error early if the test fails?

Note that even without such a test, the `Open3.popen2e` rescue would still raise an error.